server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyServerHandler.scala

```diff
@@ -40,33 +73,67 @@ class NettyServerHandler[F[_]](route: Route[F], unsafeRunAsync: (() => F[Unit])
+
+  private[this] def initHandler(ctx: ChannelHandlerContext): Unit = {
+    if (eventLoopContext == null) {
+      // Initialize our ExecutionContext
+      eventLoopContext = ExecutionContext.fromExecutor(ctx.channel.eventLoop)
+
+      // When the channel closes we want to cancel any pending dispatches.
+      // Since the listener will be executed from the channels EventLoop everything is thread safe.
+      val _ = ctx.channel.closeFuture.addListener { (_: ChannelFuture) =>
+        logger.debug(s"Http channel to ${ctx.channel.remoteAddress} closed. Cancelling ${pendingResponses.length} responses.")
+        pendingResponses.foreach(_.apply())
+      }
+    }
+  }
+
   override def channelRead0(ctx: ChannelHandlerContext, request: HttpRequest): Unit = {
 
-    def runRoute(req: HttpRequest) = {
+    def writeError500(req: HttpRequest, reason: Throwable): Unit = {
+      logger.error("Error while processing the request", reason)
+      // send 500
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.handleCloseAndKeepAliveHeaders(req)
 
-      route(NettyServerRequest(req))
-        .map {
-          case Some(response) => response
-          case None           => ServerResponse.notFound
-        }
-        .flatMap((serverResponse: ServerResponse[NettyResponse]) =>
-          // in ZIO, exceptions thrown in .map become defects - instead, we want them represented as errors so that
-          // we get the 500 response, instead of dropping the request
-          try handleResponse(ctx, req, serverResponse).unit
-          catch {
-            case e: Exception => me.error[Unit](e)
-          }
-        )
-        .handleError { case ex: Exception =>
-          logger.error("Error while processing the request", ex)
-          // send 500
-          val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
-          res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
-          res.handleCloseAndKeepAliveHeaders(req)
+      ctx.writeAndFlush(res).closeIfNeeded(req)
 
-          ctx.writeAndFlush(res).closeIfNeeded(req)
-          me.unit(())
-        }
+    }
+
+    def runRoute(req: HttpRequest, releaseReq: () => Any = () => ()): Unit = {
+      val (runningFuture, cancellationSwitch) = unsafeRunAsync { () =>
+        route(NettyServerRequest(req))
+          .map {
+            case Some(response) => response
+            case None           => ServerResponse.notFound
+          }
+      }
+      pendingResponses.enqueue(cancellationSwitch)
+      lastResponseSent = lastResponseSent.transformWith { _ =>
+        runningFuture.transform {
+          case Success(serverResponse) =>
+            pendingResponses.dequeue()
+            try {
+              handleResponse(ctx, req, serverResponse)
+              releaseReq()
```

Comment:
don't we have to run this in case of exceptions as well? the `.ensure` in the old version alway released it